node/pkg/driver/node.go

```diff
@@ -613,9 +619,82 @@ func (d *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 }
 
 func (d *NodeService) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	goid_info.SetAdditionalIDInfo(req.VolumeId)
+	volumeId := req.VolumeId
+	goid_info.SetAdditionalIDInfo(volumeId)
 	defer goid_info.DeleteAdditionalIDInfo()
-	return nil, status.Error(codes.Unimplemented, "NodeGetVolumeStats is not implemented yet")
+	volumePath := req.VolumePath
+	volumePathWithHostPrefix := d.NodeUtils.GetPodPath(volumePath)
+
+	err := d.nodeGetVolumeStatsRequestValidation(volumeId, volumePath)
+	if err != nil {
+		return nil, err
+	}
+
+	isPathExists := d.NodeUtils.IsPathExists(volumePathWithHostPrefix)
+	if !isPathExists {
+		return nil, status.Errorf(codes.NotFound, "volume path %q does not exist", volumePath)
+	}
+
+	volumeStats, err := d.getVolumeStats(volumePathWithHostPrefix, volumeId)
+	if err != nil {
+		return nil, err
+	}
+	return &csi.NodeGetVolumeStatsResponse{
+		Usage: []*csi.VolumeUsage{
+			{
+				Unit:      csi.VolumeUsage_BYTES,
+				Available: volumeStats.AvailableBytes,
+				Total:     volumeStats.TotalBytes,
+				Used:      volumeStats.UsedBytes,
+			},
+			{
+				Unit:      csi.VolumeUsage_INODES,
+				Available: volumeStats.AvailableInodes,
+				Total:     volumeStats.TotalInodes,
+				Used:      volumeStats.UsedInodes,
+			},
+		},
+	}, nil
+}
+
+func (d *NodeService) nodeGetVolumeStatsRequestValidation(volumeId string, volumePath string) error {
+	if volumeId == "" {
+		return status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume ID must be provided")
+	}
+	if volumePath == "" {
+		return status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume Path must be provided")
+	}
+
+	return nil
+}
+
+func (d *NodeService) getVolumeStats(path string, volumeId string) (VolumeStatistics, error) {
+	var volumeStats VolumeStatistics
+	errorMsg := "Failed to get statistics:"
+	isBlock, err := d.NodeUtils.IsBlock(path)
+	if err != nil {
+		return VolumeStatistics{}, status.Errorf(codes.Internal, "%q failed to determine if %q is block device: %s", errorMsg, path, err)
+	}
+
+	if isBlock {
+		mpathDevice, err := d.OsDeviceConnectivityHelper.GetMpathDevice(volumeId)
+		if err != nil {
+			switch err.(type) {
+			case *device_connectivity.MultipathDeviceNotFoundForVolumeError:
+				return VolumeStatistics{}, status.Errorf(codes.NotFound, "%q multipath Device of volume id %q does not exist", errorMsg, volumeId)
+			default:
+				return VolumeStatistics{}, status.Errorf(codes.Internal, "%q error while discovering the device : %s", errorMsg, err)
+			}
+		}
+		volumeStats, err = d.NodeUtils.GetBlockVolumeStats(mpathDevice)
+	} else {
+		volumeStats, err = d.NodeUtils.GetFileSystemVolumeStats(path)
+	}
+
+	if err != nil {
+		return VolumeStatistics{}, status.Errorf(codes.Internal, "%q failed to determine if %q is block device: %s", errorMsg, path, err)
```

Comment:
if node utils fail, it doesn't mean it failed to determine if the path is block device

Reply:
I wanted that it will tell us the function that the error was in, if you think it's not necessary I can remove it

Reply:
the error could come from `GetBlockVolumeStats`/`GetFileSystemVolumeStats`, so I think it's misleading

Reply:
ok I will remove it

Reply:
done